File: src/algorithms/arrays/stack-based/trapping-rain-water/sources/trapping-rain-water.py
# Trapping Rain Water — O(n) two-pointer approach
def trapping_rain_water(heights: list[int]) -> dict:
    array_length = len(heights)
    if array_length == 0:  # @step:initialize
        return {"total_water": 0, "water_per_index": []}  # @step:initialize

    left_pointer = 0  # @step:initialize
    right_pointer = array_length - 1  # @step:initialize
    max_left = 0  # @step:initialize
    max_right = 0  # @step:initialize
    total_water = 0  # @step:initialize
    water_per_index = [0] * array_length  # @step:initialize

    while left_pointer < right_pointer:
        if heights[left_pointer] <= heights[right_pointer]:  # @step:compare
            if heights[left_pointer] >= max_left:  # @step:compare
                max_left = heights[left_pointer]  # @step:visit
            else:
                water_per_index[left_pointer] = max_left - heights[left_pointer]  # @step:visit
                total_water += water_per_index[left_pointer]  # @step:visit
            left_pointer += 1  # @step:visit
        else:
            if heights[right_pointer] >= max_right:  # @step:compare
                max_right = heights[right_pointer]  # @step:visit
            else:
                water_per_index[right_pointer] = max_right - heights[right_pointer]  # @step:visit
                total_water += water_per_index[right_pointer]  # @step:visit
            right_pointer -= 1  # @step:visit

    return {"total_water": total_water, "water_per_index": water_per_index}  # @step:complete

File: src/algorithms/arrays/stack-based/trapping-rain-water/sources/trapping-rain-water.go
// Trapping Rain Water — O(n) two-pointer approach
package trappingrainwater

func trappingRainWater(heights []int) (totalWater int, waterPerIndex []int) {
	arrayLength := len(heights)
	if arrayLength == 0 {
		// @step:initialize
		return 0, []int{} // @step:initialize
	}

	leftPointer := 0               // @step:initialize
	rightPointer := arrayLength - 1 // @step:initialize
	maxLeft := 0                   // @step:initialize
	maxRight := 0                  // @step:initialize
	totalWater = 0                 // @step:initialize
	waterPerIndex = make([]int, arrayLength) // @step:initialize

	for leftPointer < rightPointer {
		if heights[leftPointer] <= heights[rightPointer] { // @step:compare
			if heights[leftPointer] >= maxLeft { // @step:compare
				maxLeft = heights[leftPointer] // @step:visit
			} else {
				waterPerIndex[leftPointer] = maxLeft - heights[leftPointer] // @step:visit
				totalWater += waterPerIndex[leftPointer]                    // @step:visit
			}
			leftPointer++ // @step:visit
		} else {
			if heights[rightPointer] >= maxRight { // @step:compare
				maxRight = heights[rightPointer] // @step:visit
			} else {
				waterPerIndex[rightPointer] = maxRight - heights[rightPointer] // @step:visit
				totalWater += waterPerIndex[rightPointer]                       // @step:visit
			}
			rightPointer-- // @step:visit
		}
	}

	return totalWater, waterPerIndex // @step:complete
}
